Respect dist autograd context in torch.jit._fork.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34360 Respect dist autograd context in torch.jit._fork.**

The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

pritamdamania committed:
Respect dist autograd context in torch.jit._fork.

The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

wanchaol reviewed (commented):
The approach looks good to me. But I am a bit concerned why we are keeping the `current_context_id` as thread_local? Looking at DistributedContainer I think we could make it as a static member of the DistributedContainer singleton, and this issue should just go away right?

pritamdamania87 commented:
> The approach looks good to me. But I am a bit concerned why we are keeping the current_context_id as thread_local? Looking at DistributedContainer I think we could make it as a static member of the DistributedContainer singleton, and this issue should just go away right?

Having `current_context_id` as a thread local was by design. If we have multiple JIT threads running independent forward and backward passes, each one of them needs their own unique `current_context_id` to ensure we have independent autograd graphs for each training loop for distributed autograd.

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -34,6 +37,10 @@
 namespace torch {
 namespace jit {
 
+#ifdef USE_DISTRIBUTED
```
ifdef for distributed should not appear in interpreter.cpp, this is a sign of strong coupling. I guess an abstraction where we introduce a `SavedThreadLocalState` object that holds all the pytorch-specific thread local state that interpreter needs to save and restore when it suspends. Then we have public functions that the interpreter uses in interpreter continuations:
```
unique_ptr<SavedThreadLocalState> saveThreadLocalState();
restoreThreadLocalState(const SavedThreadLocalState&);
```

ilia-cher commented:
we have `DebugInfoGuard` guard and `ThreadLocalDebugInfoBase' storage structs, their goal is to pass some information along model execution (both fwd and backward) and we also ensure that  whenever you call 'fork' or go into autograd threads we pass this info across threads too and clean up after.
So, it seems to be a good candidate in general to store this kind of data?

ilia-cher commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1249,10 +1262,16 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             // Move inputs to a separate stack
```
instead of modifying the interpreter, you could reuse, e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelThreadPoolNative.cpp#L73 and
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L515 - some update (e.g. name) is needed for ThreadLocalDebugInfoBase, since it's not going to be only for debug info;
Btw, we've just coincidentally had a discussion re. this update, and, e.g. thought about the ability to pass some arbitrary data from higher-level python layers (e.g. logical name of distributed node) too

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

pritamdamania87 commented:
Implemented @zdevito's proposal in this PR to save and restore the thread local state.

@ilia-cher I looked into the ThreadLocalDebugInfoBase structs as you suggested. Although, I felt it was probably not the best route for this change for the following reasons:

1. The API is intended for end users to use (using setThreadLocalDebugInfo) and we don't want users to mess with the thread locals for grad_mode and dist_autograd_context_id.
2. I'm not sure if we want to propagate these thread locals to all threads via at::launch. Feels safe to have them enabled only in JIT continuations for now.
3. Mixing user specified debug info and internal thread local state in a single structure didn't seem right.

pritamdamania87 commented:
cc @satgera @jjlilley

jjlilley commented on torch/csrc/ThreadLocalState.h:
```diff
@@ -0,0 +1,39 @@
+  }
+
+ private:
+  std::shared_ptr<ThreadLocalState> prev_state_;
```
Wondering aloud: Do we really need to wrap in a shared ptr, if this is [currently] just an {int64,int32}?

I understand why this makes sense (especially if we add fields in future), but it also feels expensive to add an extra malloc + free, plus several atomic ops, for each jit.fork/wait.

pritamdamania87 replied:
Good point, I'll avoid the shared_ptr for now and we can revisit this later when we add more fields.

ilia-cher commented:
> Implemented @zdevito's proposal in this PR to save and restore the thread local state.
> 
> @ilia-cher I looked into the ThreadLocalDebugInfoBase structs as you suggested. Although, I felt it was probably not the best route for this change for the following reasons:
> 
> 1. The API is intended for end users to use (using setThreadLocalDebugInfo) and we don't want users to mess with the thread locals for grad_mode and dist_autograd_context_id.
> 2. I'm not sure if we want to propagate these thread locals to all threads via at::launch. Feels safe to have them enabled only in JIT continuations for now.
> 3. Mixing user specified debug info and internal thread local state in a single structure didn't seem right.

A lot of thread local variables need to be propagated across fork and at::parallel_for, what are the examples of variables that do not need to be propagated?

ilia-cher commented:
My point is - we do want to propagate across fork and at::parallel_for and autograd threads, by default, so why not consolidate in one thread local data struct that would guarantee that properties are propagated across threads?

pritamdamania87 commented:
> A lot of thread local variables need to be propagated across fork and at::parallel_for, what are the examples of variables that do not need to be propagated?

The `dist_autograd_context_id` is only used in the forward pass and doesn't need to be propagated to autograd threads.

> My point is - we do want to propagate across fork and at::parallel_for and autograd threads, by default, so why not consolidate in one thread local data struct that would guarantee that properties are propagated across threads?

I'm a little concerned about implicitly propagating thread_locals across any arbitrary at::launch(), given that we don't know what that thread is doing and how the thread_locals might affect its behavior.

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

jjlilley reviewed (commented):
fwiw, this looks good to me now that the alloc + atomic ops are gone.
@zdevito might have more comments?

gqchen commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1147,10 +1147,9 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
                     Stack stack)
                     : state_(std::move(state)), stack_(std::move(stack)) {}
                 void operator()() {
+                  int64_t dist_autograd_context_id = -1;
```
we can remove this line?

gqchen commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1272,10 +1271,11 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             // Move inputs to a separate stack
             InterpreterState forked_interpreter(
                 frames.back().function->code_table_.at(inst.X));
+            int64_t dist_autograd_context_id = -1;
```
this line as well?

gqchen reviewed (commented):
two minor comments. Otherwise, it looks good to me.

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

zdevito reviewed (approved):
Looks good. Thanks for refactoring this!

jjlilley commented:
Pritam, this looks good to me as well, but can you hold off landing until the pending rollback lands?

We lost 500+ tests (rpc and autograd related) yesterday afternoon, the rollback seems to be landing, but would be good to test without the tests all disabled.

xush6528 commented on torch/testing/_internal/distributed/rpc/jit/dist_autograd_test.py:
```diff
@@ -3,16 +3,30 @@
+    return torch.add(t1, t2)
+
+@torch.jit.script
+def remote_add(t1, t2, dst: str):
```
To silent linter, need to annotate `def remote_add(t1, t2, dst: str):  # noqa: E999`

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]

pritamdamania87 commented:
As per Jeremy's suggestion, will hold off on landing this until things stabilize in trunk.

pritamdamania committed:
Update on "Respect dist autograd context in torch.jit._fork."


The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

[ghstack-poisoned]
